### PR TITLE
Add unique index to DedupeRuleGroup.name field

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Ensures DedupeRuleGroup.name field is always unique by adding a db constraint.

Before
----------------------------------------
Uniqueness was sorta ensured at the form layer.

After
----------------------------------------
Centralized logic from #22570 plus a unique constraint ensures that it's always unique. Form layer creativity removed.

Comments
----------------------------------------
Yea... there's kind of no point having a "name" field at all if we don't enforce uniqueness.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I agree with the sentiment - but some work on tests 

[Test Result](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/) (18 failures / +18)
[CRM_Contact_Import_Parser_ContactTest.testImportParserWithUpdateWithCustomRuleNoExternalIDMatch](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/CRM_Contact_Import_Parser_ContactTest/testImportParserWithUpdateWithCustomRuleNoExternalIDMatch/)
[CRM_Contact_Import_Parser_ContactTest.testIgnoreLocationTypeId](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/CRM_Contact_Import_Parser_ContactTest/testIgnoreLocationTypeId/)
[api_v3_ContactTest.testGetMatchesPhoneWithType with data set #1](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/api_v3_ContactTest/testGetMatchesPhoneWithType_with_data_set__1/)
[api_v3_ContactTest.testGetMatchesPhoneWithType with data set #2](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/api_v3_ContactTest/testGetMatchesPhoneWithType_with_data_set__2/)
[api_v3_JobTest.testBatchMergeEmptyRule with data set #3](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/api_v3_JobTest/testBatchMergeEmptyRule_with_data_set__3/)
[api_v3_JobTest.testBatchMergeEmptyRule with data set #4](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/api_v3_JobTest/testBatchMergeEmptyRule_with_data_set__4/)
[api_v3_JobTest.testBatchMergeEmptyRule with data set #8](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/api_v3_JobTest/testBatchMergeEmptyRule_with_data_set__8/)
[api_v3_JobTest.testBatchMergeEmptyRule with data set #9](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/api_v3_JobTest/testBatchMergeEmptyRule_with_data_set__9/)
[api_v3_JobTest.testBatchMergeEmptyRule with data set #13](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/api_v3_JobTest/testBatchMergeEmptyRule_with_data_set__13/)
[api_v3_JobTest.testBatchMergeEmptyRule with data set #14](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/testReport/junit/(root)/api_v3_JobTest/testBatchMergeEmptyRule_with_data_set__14/)
[Show all failed tests >>>](https://test.civicrm.org/job/CiviCRM-Core-PR/51079/#showFailuresLink)

**Comment by colemanw** [comment]
All failures were due to test setup creating duplicate rule names. Fixed.

File: CRM/Upgrade/Incremental/php/FiveFiftyFour.php [hunk 1]
```diff
@@ -50,6 +52,17 @@ public static function addCreatedIDColumnToParticipant($ctx): bool {
...
   }
 
+  /**
+   * @param \CRM_Queue_TaskContext $ctx
+   *
+   * @return bool
+   */
+  public static function alterDedupeRuleGroupName(CRM_Queue_TaskContext $ctx) {
+    CRM_Core_DAO::executeQuery("ALTER TABLE `civicrm_dedupe_rule_group` CHANGE COLUMN `name` `name` varchar(255) COMMENT 'Unique name of rule group'", [], TRUE, NULL, FALSE, FALSE);
+    CRM_Core_DAO::executeQuery("UPDATE `civicrm_dedupe_rule_group` g1, `civicrm_dedupe_rule_group` g2 SET g1.name = CONCAT(g1.name, '_', g1.id) WHERE g1.name = g2.name AND g1.id > g2.id", [], TRUE, NULL, FALSE, FALSE);
```
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/php/FiveFiftyFour.php:62]: I think you need to reverse the order of these two statements

Comment (colemanw) [reply CRM/Upgrade/Incremental/php/FiveFiftyFour.php:62]: @eileenmcnaughton why? The first one just makes the column longer, which makes room for the suffix appended in the 2nd statement.

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveFiftyFour.php:62]: Also I think the id > will be safe in practice - but we don't want to rename anything that `is_reserved` as the name has meaning then

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveFiftyFour.php:62]: Actually no - the issue is you aren't creating the indexes - CRM_Core_BAO_SchemaHandler::createIndexes('civicrm_dedupe_rule_group', ['name']);

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveFiftyFour.php:62]: hmm - maybe you are but not here...

File: CRM/Upgrade/Incremental/php/FiveFiftyFour.php [hunk 2]
```diff
@@ -39,6 +39,8 @@ public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NU
   public function upgrade_5_54_alpha1($rev): void {
     $this->addTask(ts('Upgrade DB to %1: SQL', [1 => $rev]), 'runSql', $rev);
     $this->addTask('Add "created_id" column to "civicrm_participant"', 'addCreatedIDColumnToParticipant');
+    $this->addTask('Increase field length of civicrm_dedupe_rule_group.name', 'alterDedupeRuleGroupName');
+    $this->addTask('Add index civicrm_dedupe_rule_group.UI_name', 'addIndex', 'civicrm_dedupe_rule_group', 'name', 'UI');
```
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/php/FiveFiftyFour.php:43]: ah

**Comment by eileenmcnaughton** [comment]
OK - I've been my own ruber duck - all good, merging

**Comment by seamuslee001** [comment]
@colemanw I believe this was in reference to this https://rubberduckdebugging.com/ but Eileen suggesting she was playing the role of the duck as well

**Comment by eileenmcnaughton** [comment]
quack

**Pull request merged**